util/fibers/future.h

```diff
@@ -4,339 +4,39 @@
-
-  promise_base() : promise_base{std::allocator_arg, std::allocator<promise_base>{}} {
-  }
-
-  template <typename Allocator> promise_base(std::allocator_arg_t, Allocator alloc) {
-    typedef detail::shared_state_object<R, Allocator> object_type;
-    typedef std::allocator_traits<typename object_type::allocator_type> traits_type;
-    typename object_type::allocator_type a{alloc};
-    typename traits_type::pointer ptr{traits_type::allocate(a, 1)};
-    auto* p = ptr;
-
-    try {
-      traits_type::construct(a, p, a);
-    } catch (...) {
-      traits_type::deallocate(a, ptr, 1);
-      throw;
-    }
-    future_.reset(p);
-  }
-
-  ~promise_base() {
-    if (future_ && obtained_) {
-      // future_->owner_destroyed();
-    }
-  }
-
-  promise_base(promise_base const&) = delete;
-  promise_base& operator=(promise_base const&) = delete;
-
-  promise_base(promise_base&& other) noexcept
-      : obtained_{other.obtained_}, future_{std::move(other.future_)} {
-    other.obtained_ = false;
-  }
-
-  promise_base& operator=(promise_base&& other) noexcept {
-    if ((this != &other)) {
-      promise_base tmp{std::move(other)};
-      swap(tmp);
-    }
-    return *this;
-  }
-
-  Future<R> get_future() {
-    assert(!obtained_);
-    assert(future_);
-    obtained_ = true;
-    return Future<R>{future_};
-  }
-
-  void swap(promise_base& other) noexcept {
-    std::swap(obtained_, other.obtained_);
-    future_.swap(other.future_);
-  }
-};
-
-}  // namespace detail
-
-template <typename R> class Future : private detail::future_base<R> {
- private:
-  typedef detail::future_base<R> base_type;
-
-  friend struct detail::promise_base<R>;
-
-  explicit Future(typename base_type::ptr_type const& p) noexcept : base_type{p} {
-  }
-
- public:
-  Future() = default;
-
-  Future(Future const&) = delete;
-  Future& operator=(Future const&) = delete;
-
-  Future(Future&& other) noexcept : base_type{std::move(other)} {
-  }
-
-  Future& operator=(Future&& other) noexcept {
-    if ((this != &other)) {
-      base_type::operator=(std::move(other));
-    }
-    return *this;
-  }
-
-  R get() {
-    typename base_type::ptr_type tmp{};
-    tmp.swap(base_type::state_);
-    return std::move(tmp->get());
-  }
-
-  using base_type::valid;
-  using base_type::wait;
-  using base_type::wait_for;
-  using base_type::wait_until;
-};
-
-template <typename R> class Promise : private detail::promise_base<R> {
-  typedef detail::promise_base<R> base_type;
-
- public:
-  void set_value(R& value) {
-    base_type::future_->set_value(value);
-  }
-
-  void set_value(const R& value) {
-    base_type::future_->set_value(value);
-  }
-
-  void swap(Promise& other) noexcept {
-    base_type::swap(other);
-  }
-
-  using base_type::get_future;
+  struct Block {
+    T value{};  // replace with aligned_storage or optional if T is not default-constructible
+    std::atomic_bool has_value = false;
+    util::fb2::EventCount waker{};
+  };
+  std::shared_ptr<Block> block;
```

Comment:
if you extend to  `std::variant<std::shared_ptr<Block>, T>` you could provide support for immediate future built-in.

Reply:
I already have StringValue in the main repo